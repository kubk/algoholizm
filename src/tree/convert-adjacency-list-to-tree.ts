export type NodeId = string | number | null;

export type Flat<T extends NodeId> = Array<{
  id: T;
  name: string;
  parentId: T;
}>;

export type Tree<T extends NodeId> = {
  id: T;
  name: string;
  children: Array<Tree<T>>;
};

export const convertAdjacencyListToTree = <T extends NodeId>(
  startFromId: T,
  flatList: Flat<T>
): Array<Tree<T>> => {
  return flatList
    .filter((node) => node.parentId === startFromId)
    .map((rootNode) => ({
      id: rootNode.id,
      name: rootNode.name,
      children: convertAdjacencyListToTree(rootNode.id, flatList),
    }));
};
