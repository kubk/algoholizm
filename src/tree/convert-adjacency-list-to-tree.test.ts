import { convertAdjacencyListToTree } from './convert-adjacency-list-to-tree';
import { describe, expect, it } from 'vitest';

describe('convertAdjacencyListToTree', () => {
  it('converts number ids', () => {
    const flat = [
      { id: 1, name: 'Phonedo', parentId: 0 },
      { id: 2, name: 'Phonedo - A', parentId: 1 },
      { id: 3, name: 'Phonedo - A - A', parentId: 2 },
      { id: 4, name: 'Phonedo - B', parentId: 1 },
      { id: 6, name: 'Phonedo - C', parentId: 1 },
      { id: 7, name: 'Phonedo - A - B', parentId: 2 },
    ];

    expect(convertAdjacencyListToTree(0, flat)).toMatchSnapshot();
  });

  it('converts null|string ids', () => {
    const flat = [
      { id: '1', name: 'Phonedo', parentId: null },
      { id: '2', name: 'Phonedo - A', parentId: '1' },
      { id: '3', name: 'Phonedo - A - A', parentId: '2' },
      { id: '4', name: 'Phonedo - B', parentId: '1' },
      { id: '6', name: 'Phonedo - C', parentId: '1' },
      { id: '7', name: 'Phonedo - A - B', parentId: '2' },
    ];

    expect(convertAdjacencyListToTree(null, flat)).toMatchSnapshot();
  });
});
