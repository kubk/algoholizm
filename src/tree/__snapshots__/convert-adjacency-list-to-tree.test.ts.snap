// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertAdjacencyListToTree > converts null|string ids 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "3",
            "name": "Phonedo - A - A",
          },
          {
            "children": [],
            "id": "7",
            "name": "Phonedo - A - B",
          },
        ],
        "id": "2",
        "name": "Phonedo - A",
      },
      {
        "children": [],
        "id": "4",
        "name": "Phonedo - B",
      },
      {
        "children": [],
        "id": "6",
        "name": "Phonedo - C",
      },
    ],
    "id": "1",
    "name": "Phonedo",
  },
]
`;

exports[`convertAdjacencyListToTree > converts number ids 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": 3,
            "name": "Phonedo - A - A",
          },
          {
            "children": [],
            "id": 7,
            "name": "Phonedo - A - B",
          },
        ],
        "id": 2,
        "name": "Phonedo - A",
      },
      {
        "children": [],
        "id": 4,
        "name": "Phonedo - B",
      },
      {
        "children": [],
        "id": 6,
        "name": "Phonedo - C",
      },
    ],
    "id": 1,
    "name": "Phonedo",
  },
]
`;

exports[`convertAdjacencyListToTree converts null|string ids 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": "3",
            "name": "Phonedo - A - A",
          },
          Object {
            "children": Array [],
            "id": "7",
            "name": "Phonedo - A - B",
          },
        ],
        "id": "2",
        "name": "Phonedo - A",
      },
      Object {
        "children": Array [],
        "id": "4",
        "name": "Phonedo - B",
      },
      Object {
        "children": Array [],
        "id": "6",
        "name": "Phonedo - C",
      },
    ],
    "id": "1",
    "name": "Phonedo",
  },
]
`;

exports[`convertAdjacencyListToTree converts number ids 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "id": 3,
            "name": "Phonedo - A - A",
          },
          Object {
            "children": Array [],
            "id": 7,
            "name": "Phonedo - A - B",
          },
        ],
        "id": 2,
        "name": "Phonedo - A",
      },
      Object {
        "children": Array [],
        "id": 4,
        "name": "Phonedo - B",
      },
      Object {
        "children": Array [],
        "id": 6,
        "name": "Phonedo - C",
      },
    ],
    "id": 1,
    "name": "Phonedo",
  },
]
`;
